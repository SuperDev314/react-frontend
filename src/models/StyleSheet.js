// @flow

import { cloneElement } from 'react'
import {
  IS_BROWSER,
  DISABLE_SPEEDY,
  SC_ATTR,
  SC_STREAM_ATTR,
} from '../constants'
import { makeTag, makeRehydrationTag, type Tag } from './StyleTags'
import extractComps from '../utils/extractCompsFromCSS'

/* determine the maximum number of components before tags are sharded */
let MAX_SIZE
if (IS_BROWSER) {
  /* in speedy mode we can keep a lot more rules in a sheet before a slowdown can be expected */
  MAX_SIZE = DISABLE_SPEEDY ? 40 : 1000
} else {
  /* for servers we do not need to shard at all */
  MAX_SIZE = -1
}

let sheetRunningId = 0
let global

class StyleSheet {
  id: number
  sealed: boolean
  forceServer: boolean
  target: ?HTMLElement
  tagMap: { [string]: Tag<any> }
  hashes: { [string]: string }
  deferred: { [string]: string[] }
  rehydratedNames: { [string]: boolean }
  tags: Tag<any>[]
  capacity: number
  clones: StyleSheet[]

  constructor(
    target: ?HTMLElement = IS_BROWSER ? document.head : null,
    forceServer?: boolean = false
  ) {
    this.id = sheetRunningId += 1
    this.sealed = false
    this.forceServer = forceServer
    this.target = forceServer ? null : target
    this.tagMap = {}
    this.hashes = {}
    this.deferred = {}
    this.rehydratedNames = {}
    this.tags = []
    this.capacity = 1
    this.clones = []
  }

  /* rehydrate all SSR'd style tags */
  rehydrate() {
    if (!IS_BROWSER || this.forceServer) {
      return this
    }

    const els = []
    const names = []
    let extracted = []
    let isStreamed = false

    /* retrieve all of our SSR style elements from the DOM */
    const nodes = document.querySelectorAll(`style[${SC_ATTR}]`)
    const nodesSize = nodes.length

    /* abort rehydration if no previous style tags were found */
    if (nodesSize === 0) {
      return this
    }

    for (let i = 0; i < nodesSize; i += 1) {
      // $FlowFixMe: We can trust that all elements in this query are style elements
      const el = (nodes[i]: HTMLStyleElement)

      /* check if style tag is a streamed tag */
      isStreamed = !!el.getAttribute(SC_STREAM_ATTR) || isStreamed

      /* retrieve all component names */
      const elNames = (el.getAttribute(SC_ATTR) || '').trim().split(/\s+/)
      const elNamesSize = elNames.length
      for (let j = 0; j < elNamesSize; j += 1) {
        const name = elNames[j]
        /* add rehydrated name to sheet to avoid readding styles */
        this.rehydratedNames[name] = true
        names.push(name)
      }

      /* extract all components and their CSS */
      extracted = extracted.concat(extractComps(el.textContent))
      /* store original HTMLStyleElement */
      els.push(el)
    }

    /* abort rehydration if nothing was extracted */
    const extractedSize = extracted.length
    if (extractedSize === 0) {
      return this
    }

    /* create a tag to be used for rehydration */
    const tag = makeTag(this.target, null, this.forceServer)
    const rehydrationTag = makeRehydrationTag(
      tag,
      els,
      extracted,
      names,
      isStreamed
    )

    /* reset capacity and adjust MAX_SIZE by the initial size of the rehydration */
    this.capacity = Math.max(1, MAX_SIZE - extractedSize)
    this.tags.push(rehydrationTag)

    /* retrieve all component ids */
    for (let j = 0; j < extractedSize; j += 1) {
      this.tagMap[extracted[j].componentId] = rehydrationTag
    }

    return this
  }

  /* retrieve a "global" instance of StyleSheet which is typically used when no other is available */
  static get global(): StyleSheet {
    return global || (global = new StyleSheet().rehydrate())
  }

  /* NOTE: This is just for backwards-compatibility with jest-styled-components */
  static get instance(): StyleSheet {
    return StyleSheet.global
  }

  /* reset the internal "global" instance */
  static reset(forceServer?: boolean = false) {
    global = new StyleSheet(undefined, forceServer).rehydrate()
  }

  /* adds "children" to the StyleSheet that inherit all of the parents' rules
   * while their own rules do not affect the parent */
  clone() {
    const sheet = new StyleSheet(this.target, this.forceServer)
    /* add to clone array */
    this.clones.push(sheet)

    /* clone all tags */
    sheet.tags = this.tags.map(tag => {
      const ids = tag.getIds()
      const newTag = tag.clone()

      /* reconstruct tagMap */
      for (let i = 0; i < ids.length; i += 1) {
        sheet.tagMap[ids[i]] = newTag
      }

      return newTag
    })

    /* clone other maps */
    sheet.rehydratedNames = { ...this.rehydratedNames }
    sheet.deferred = { ...this.deferred }
    sheet.hashes = { ...this.hashes }

    return sheet
  }

  /* force StyleSheet to create a new tag on the next injection */
  sealAllTags() {
    this.capacity = 1
    this.sealed = true
  }

  /* get a tag for a given componentId, assign the componentId to one, or shard */
  getTagForId(id: string): Tag<any> {
    /* simply return a tag, when the componentId was already assigned one */
    const prev = this.tagMap[id]
    if (prev !== undefined && !this.sealed) {
      return prev
    }

    let tag = this.tags[this.tags.length - 1]

    /* shard (create a new tag) if the tag is exhausted (See MAX_SIZE) */
    this.capacity -= 1
    if (this.capacity === 0) {
      this.capacity = MAX_SIZE
      this.sealed = false
      tag = makeTag(this.target, tag ? tag.styleTag : null, this.forceServer)
      this.tags.push(tag)
    }

    return (this.tagMap[id] = tag)
  }

  /* optimal caching: hash is known and name can be returned */
  getNameForHash(hash: string) {
    return this.hashes[hash]
  }

  /* rehydration check: name is known, hash is unknown, but styles are present */
  alreadyInjected(hash: any, name: string) {
    if (!this.rehydratedNames[name]) return false

    this.hashes[hash] = name
    return true
  }

  /* checks whether component is already registered */
  hasInjectedComponent(id: string): boolean {
    return !!this.tagMap[id]
  }

  /* registers a componentId and registers it on its tag */
  deferredInject(id: string, cssRules: string[]) {
    const { clones } = this
    for (let i = 0; i < clones.length; i += 1) {
      clones[i].deferredInject(id, cssRules)
    }

    this.getTagForId(id).insertMarker(id)
    this.deferred[id] = cssRules
  }

  inject(id: string, cssRules: string[], hash?: string, name?: string) {
    const { clones } = this
    for (let i = 0; i < clones.length; i += 1) {
      clones[i].inject(id, cssRules, hash, name)
    }

    /* add deferred rules for component */
    let injectRules = cssRules
    const deferredRules = this.deferred[id]
    if (deferredRules !== undefined) {
      injectRules = deferredRules.concat(injectRules)
      delete this.deferred[id]
    }

    const tag = this.getTagForId(id)
    tag.insertRules(id, injectRules)

    if (hash && name) {
      tag.names.push(name)
      this.hashes[hash] = name
    }
  }

  toHTML() {
    return this.tags.map(tag => tag.toHTML()).join('')
  }

  toReactElements() {
    const { id } = this

    return this.tags.map((tag, i) => {
      const key = `sc-${id}-${i}`
      return cloneElement(tag.toElement(), { key })
    })
  }
}

export default StyleSheet
