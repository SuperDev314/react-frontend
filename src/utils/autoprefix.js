import camelizeStyleName from 'fbjs/lib/camelizeStyleName'
import hyphenateStyleName from 'fbjs/lib/hyphenateStyleName'
import { autoprefix } from 'glamor/lib/autoprefix'

export default root => {
  root.walkDecls(decl => {
    const objStyle = { [camelizeStyleName(decl.prop)]: decl.value }
    const prefixed = autoprefix(objStyle)
    Object.keys(prefixed).reverse().forEach(newProp => {
      decl.cloneBefore({
        prop: hyphenateStyleName(newProp),
        value: prefixed[newProp],
      })
    })
    decl.remove()
  })
}
