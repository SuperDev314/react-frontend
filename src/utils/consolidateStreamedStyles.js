// @flow
/**
 * When using streaming rendering, style blocks are emitted in chunks directly.
 * This method used to be necessary to move all style elements before the
 * rehydration kicks in. However the rehydration now takes all SSR style tags,
 * combines them into a single tag, and adds it to the head.
 * This renders the consolidation useless
 */
export default function consolidateStreamedStyles() {
  /* noop */
}
