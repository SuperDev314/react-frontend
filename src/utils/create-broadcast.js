/**
 * Creates a broadcast that can be listened to, i.e. simple event emitter
 *
 * @see https://github.com/ReactTraining/react-broadcast
 */
const createBroadcast = (initialValue) => {
  let listeners = []
  let currentValue = initialValue

  return {
    publish(value) {
      currentValue = value
      listeners.forEach(listener => listener(currentValue))
    },
    subscribe(listener) {
      listeners.push(listener)

      // Publish to this subscriber once immediately.
      listener(currentValue)

      // eslint-disable-next-line no-return-assign
      return () =>
        listeners = listeners.filter(item => item !== listener)
    },
  }
}

export default createBroadcast
