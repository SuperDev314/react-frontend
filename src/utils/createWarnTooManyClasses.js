// @flow

import styleSheet from '../models/StyleSheet'

const LIMIT = 200

export default () => {
  let generatedClasses = {}
  let warningSeen = false

  return (className: string, displayName: string) => {
    if (!warningSeen) {
      generatedClasses[className] = true
      if (Object.keys(generatedClasses).length >= LIMIT) {
        const latestRule = styleSheet.rules().find(s => s.selectorText && s.selectorText.indexOf(`.${className}`) !== -1)
        // Unable to find latestRule in test environment.
        /* eslint-disable no-console, prefer-template */
        console.warn(`Over ${LIMIT} classes were generated for component ${displayName}. ` +
           'Consider using style property for frequently changed styles.\n' +
           'Example: const MyStyledComponent = styled(props => <div className={props.className} style={{color: props.color}}>{props.children}</div>)`width: 100px; height: 100px;`' +
           (latestRule ? `\nLatest generated class: ${latestRule.cssText}` : ''))
        warningSeen = true
        generatedClasses = {}
      }
    }
  }
}
