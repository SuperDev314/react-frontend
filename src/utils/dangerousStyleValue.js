// @flow
import unitless from '@emotion/unitless';

// Taken from https://github.com/facebook/react/blob/b87aabdfe1b7461e7331abb3601d9e6bb27544bc/packages/react-dom/src/shared/dangerousStyleValue.js
export default function dangerousStyleValue(name: string, value: any): any {
  // https://github.com/amilajack/eslint-plugin-flowtype-errors/issues/133
  // $FlowFixMe
  if (typeof value === 'symbol') {
    return value;
  }

  const isEmpty = value == null || typeof value === 'boolean' || value === '';
  if (isEmpty) {
    return '';
  }

  if (
    typeof value === 'number' &&
    value !== 0 &&
    !(unitless.hasOwnProperty(name) && unitless[name] === 1)
  ) {
    return `${value}px`; // Presumes implicit 'px' suffix for unitless numbers
  }

  return String(value).trim();
}
