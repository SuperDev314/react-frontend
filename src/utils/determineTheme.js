// @flow
import { EMPTY_OBJECT } from './empties';

type Props = {
  theme?: any,
};

export default (props: Props, fallbackTheme: any, defaultProps: any = EMPTY_OBJECT) => {
  // Props should take precedence over ThemeProvider, which should take precedence over
  // defaultProps, but React automatically puts defaultProps on props.

  /* eslint-disable react/prop-types, flowtype-errors/show-errors */
  const isDefaultTheme = defaultProps ? props.theme === defaultProps.theme : false;
  const theme = props.theme && !isDefaultTheme ? props.theme : fallbackTheme || defaultProps.theme;
  /* eslint-enable */

  return theme;
};
