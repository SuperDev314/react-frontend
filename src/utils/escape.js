// @flow
const escapeRegex = /[[\].#*$><+~=|^:(),"'`]/g
const multiDashRegex = /--+/g
const dashAtEnd = /-$/

export default function escape(str: string): string {
  return str
        // Replace all possible CSS selectors
        .replace(escapeRegex, '-')

        // Replace multiple -- with single -
        .replace(multiDashRegex, '-')

        // Remove extraneous hyphens at the end
        .replace(dashAtEnd, '')
}
