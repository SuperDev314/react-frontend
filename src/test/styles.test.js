// @flow
import React from 'react';
import TestRenderer from 'react-test-renderer';

import * as nonce from '../utils/nonce';
import { resetStyled, expectCSSMatches } from './utils';
import StyleSheet from '../models/StyleSheet';

jest.mock('../utils/nonce');
jest.spyOn(nonce, 'default').mockImplementation(() => 'foo');

let styled;

describe('with styles', () => {
  /**
   * Make sure the setup is the same for every test
   */
  beforeEach(() => {
    // $FlowFixMe
    document.head.innerHTML = '';
    styled = resetStyled();
  });

  it('should append a style', () => {
    const rule = 'color: blue;';
    const Comp = styled.div`
      ${rule};
    `;
    TestRenderer.create(<Comp />);
    expectCSSMatches('.b { color:blue; }');
  });

  it('should append multiple styles', () => {
    const rule1 = 'color: blue;';
    const rule2 = 'background: red;';
    const Comp = styled.div`
      ${rule1} ${rule2};
    `;
    TestRenderer.create(<Comp />);
    expectCSSMatches('.b { color:blue; background:red; }');
  });

  it('amperstand should refer to the static class when making a self-referential combo selector', () => {
    const Comp = styled.div`
      background: red;
      color: ${p => p.color};

      &&& {
        border: 1px solid red;
      }

      &[disabled] {
        color: red;

        & + & {
          margin-bottom: 4px;
        }

        & > & {
          margin-top: 4px;
        }
      }

      & + & {
        margin-left: 4px;
      }

      & + & ~ & {
        background: black;
      }

      & ~ & {
        margin-right: 4px;
      }

      & > & {
        margin-top: 4px;
      }

      .foo & {
        color: silver;
      }

      .foo > & {
        color: green;
      }
    `;
    TestRenderer.create(
      <React.Fragment>
        <Comp color="white" />
        <Comp color="red" />
      </React.Fragment>
    );
    expectCSSMatches(`
      .b{ background:red; color:white; }
      .b.b.b{ border:1px solid red; }
      .b[disabled]{ color:red; }
      .b[disabled] + .sc-a[disabled]{ margin-bottom:4px; }
      .b[disabled] > .sc-a[disabled]{ margin-top:4px; }
      .b + .sc-a{ margin-left:4px; }
      .b + .sc-a ~ .sc-a{ background:black; }
      .b ~ .sc-a{ margin-right:4px; }
      .b > .sc-a{ margin-top:4px; }
      .foo .b{ color:silver; }
      .foo > .b{ color:green; }
      .c{ background:red; color:red; }
      .c.c.c{ border:1px solid red; }
      .c[disabled]{ color:red; }
      .c[disabled] + .sc-a[disabled]{ margin-bottom:4px; }
      .c[disabled] > .sc-a[disabled]{ margin-top:4px; }
      .c + .sc-a{ margin-left:4px; }
      .c + .sc-a ~ .sc-a{ background:black; }
      .c ~ .sc-a{ margin-right:4px; }
      .c > .sc-a{ margin-top:4px; }
      .foo .c{ color:silver; }
      .foo > .c{ color:green; }
    `);
  });

  it('should handle inline style objects', () => {
    const rule1 = {
      backgroundColor: 'blue',
    };
    const Comp = styled.div`
      ${rule1};
    `;
    TestRenderer.create(<Comp />);
    expectCSSMatches('.b { background-color:blue; }');
  });

  it('should handle inline style objects with media queries', () => {
    const rule1 = {
      backgroundColor: 'blue',
      '@media screen and (min-width: 250px)': {
        backgroundColor: 'red',
      },
    };
    const Comp = styled.div`
      ${rule1};
    `;
    TestRenderer.create(<Comp />);
    expectCSSMatches(
      '.b { background-color:blue; } @media screen and (min-width:250px) { .b { background-color:red; } }'
    );
  });

  it('should handle inline style objects with pseudo selectors', () => {
    const rule1 = {
      backgroundColor: 'blue',
      '&:hover': {
        color: 'green',
      },
    };
    const Comp = styled.div`
      ${rule1};
    `;
    TestRenderer.create(<Comp />);
    expectCSSMatches('.b { background-color:blue; } .b:hover { color:green; }');
  });

  it('should handle inline style objects with pseudo selectors', () => {
    const rule1 = {
      backgroundColor: 'blue',
      '&:hover': {
        color: 'green',
      },
    };
    const Comp = styled.div`
      ${rule1};
    `;
    TestRenderer.create(<Comp />);
    expectCSSMatches('.b { background-color:blue; } .b:hover { color:green; }');
  });

  it('should handle inline style objects with nesting', () => {
    const rule1 = {
      backgroundColor: 'blue',
      '> h1': {
        color: 'white',
      },
    };
    const Comp = styled.div`
      ${rule1};
    `;
    TestRenderer.create(<Comp />);
    expectCSSMatches('.b { background-color:blue; } .b > h1 { color:white; }');
  });

  it('should handle inline style objects with contextual selectors', () => {
    const rule1 = {
      backgroundColor: 'blue',
      'html.something &': {
        color: 'white',
      },
    };
    const Comp = styled.div`
      ${rule1};
    `;
    TestRenderer.create(<Comp />);
    expectCSSMatches('.b { background-color:blue; } html.something .b { color:white; }');
  });

  it('should inject styles of multiple components', () => {
    const firstRule = 'background: blue;';
    const secondRule = 'background: red;';
    const FirstComp = styled.div`
      ${firstRule};
    `;
    const SecondComp = styled.div`
      ${secondRule};
    `;

    TestRenderer.create(<FirstComp />);
    TestRenderer.create(<SecondComp />);

    expectCSSMatches('.c { background:blue; } .d { background:red; }');
  });

  it('should inject styles of multiple components based on creation, not rendering order', () => {
    const firstRule = 'content: "first rule";';
    const secondRule = 'content: "second rule";';
    const FirstComp = styled.div`
      ${firstRule};
    `;
    const SecondComp = styled.div`
      ${secondRule};
    `;

    // Switch rendering order, shouldn't change injection order
    TestRenderer.create(<SecondComp />);
    TestRenderer.create(<FirstComp />);

    // Classes _do_ get generated in the order of rendering but that's ok
    expectCSSMatches(`
        .d { content:"first rule"; }
        .c { content:"second rule"; }
      `);
  });

  it('should strip a JS-style (invalid) comment in the styles', () => {
    const comment = '// This is an invalid comment';
    const rule = 'color: blue;';
    // prettier-ignore
    const Comp = styled.div`
      ${comment}
      ${rule}
    `
    TestRenderer.create(<Comp />);
    expectCSSMatches(`
        .b {
          color:blue;
        }
      `);
  });

  it('should respect removed rules', () => {
    const Heading = styled.h1`
      color: red;
    `;
    const Text = styled.span`
      color: green;
    `;

    TestRenderer.create(
      <Heading>
        <Text />
      </Heading>
    );
    StyleSheet.master.remove(Text.styledComponentId);

    expectCSSMatches(`
        .c {
          color:red;
        }
      `);
  });

  it('should add a webpack nonce to the style tags if one is available in the global scope', () => {
    const rule = 'color: blue;';
    const Comp = styled.div`
      ${rule};
    `;
    TestRenderer.create(<Comp />);
    expectCSSMatches(`
        .b {
          color:blue;
        }
      `);

    Array.from(document.querySelectorAll('style')).forEach(el => {
      expect(el.getAttribute('nonce')).toBe('foo');
    });
  });

  it('should handle deferredInject and inject correctly', () => {
    const cloneA = StyleSheet.master.clone();
    const cloneB = StyleSheet.master.clone();
    const rules = ['.testA {}'];

    StyleSheet.master.deferredInject('test', rules);

    expect(StyleSheet.master.deferred.test).toBe(rules);
    expect(cloneA.deferred.test).toBe(rules);
    expect(cloneB.deferred.test).toBe(rules);

    StyleSheet.master.inject('test', ['.testB {}']);

    const inspectTag = sheet => {
      const tag = sheet.getTagForId('test');
      return tag.css();
    };

    const masterCss = inspectTag(StyleSheet.master);

    expect(masterCss).toEqual(inspectTag(cloneA));
    expect(masterCss).toEqual(inspectTag(cloneB));
  });
});
