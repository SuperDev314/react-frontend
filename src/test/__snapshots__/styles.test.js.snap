// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with styles should throw a meaningful error if a non styled component with react element as root child is referred as a nested style 1`] = `"A plain React element (NestedComp) has been interpolated into styles, probably as a component selector (https://www.styled-components.com/docs/advanced#referring-to-other-components). Only styled-component classes can be targeted in this fashion."`;
